lib/wasi/src/state/pipe.rs

```diff
@@ -235,18 +235,27 @@ impl WasiPipe {
         iov: WasmSlice<__wasi_iovec_t<M>>,
     ) -> Result<usize, Errno> {
         loop {
-            if let Some(buf) = self.read_buffer.as_mut() {
-                let buf_len = buf.len();
-                if buf_len > 0 {
-                    let reader = buf.as_ref();
-                    let read = read_bytes(reader, memory, iov).map(|_| buf_len as usize)?;
-                    buf.advance(read);
-                    return Ok(read);
+            if let Ok(mut read_buffer) = self.read_buffer.lock() {
+                if let Some(buf) = read_buffer.as_mut() {
+                    let buf_len = buf.len();
+                    if buf_len > 0 {
+                        let reader = buf.as_ref();
+                        let read = read_bytes(reader, memory, iov).map(|_| buf_len as usize)?;
```

Comment:
Seems like you are throwing away the actual amount of bytes read, which can be lower than the buffer size.